haas/cli.py

```diff
@@ -31,8 +32,55 @@
 MIN_PORT_NUMBER = 1
 MAX_PORT_NUMBER = 2**16 - 1
 
+
+class HTTPClient(object):
+    """An HTTP client.
+
+    Makes HTTP requests on behalf of the HaaS CLI. Responsible for adding
+    authentication information to the request.
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def request(method, url, data=None):
+        """Make an HTTP request
+
+        Makes an HTTP request on URL `url` with method `method` and request
+        body `data` (if supplied). May add authentication or other
+        backend-specific information to the request.
+
+        Parameters
+        ----------
+
+        method : str
+            The HTTP method to use, e.g. 'GET', 'PUT', 'POST'...
+        url : str
+            The URL to act on
+        data : str, optional
+            The body of the request
+
+        Returns
+        -------
+
+        requests.Response
+            The HTTP response
+        """
+
+
+class RequestsHTTPClient(requests.Session, HTTPClient):
```

Comment:
Quoting Kristi Nikolla (2016-07-20 13:13:56)

> > +class RequestsHTTPClient(requests.Session, HTTPClient):
>
>    I understand the reasoning behind the wrapper, however why not just use
>    the plain requests.Session for the fallback case? It will still work
>    through duck typing.

I don't have a strong opinion, but it does seem like we should document
somewhere explicitly that it satisfies the interface, and this seemed as
good a place as any. If folks have a preference for not wrapping
requests.Session, I'm happy to make that change. I'll wait to hear from
@henn and @gsilvis first though.